torch/fx/passes/runtime_assert.py

```diff
@@ -85,60 +93,140 @@ def insert_deferred_runtime_asserts(
         free_symbols,
         InnerTensorKey,
     )
-    from torch.utils._sympy.interp import sympy_interp
     from torch.utils._sympy.numbers import int_oo
     from torch.utils._sympy.reference import PythonReferenceAnalysis
 
     # TODO: Request simplification on runtime asserts before emitting them
     ras_by_symbol = shape_env.deferred_runtime_asserts.copy()
     graph = gm.graph
-
-    if not any(ras for ras in ras_by_symbol.values()):
-        return
-
     graph_code_log.debug(
         "%s",
         lazy_format_graph_code(
             f"pre insert_deferred_runtime_asserts {name}", gm, colored=True
         ),
     )
 
-    # deduplicate unassociated runtime assertions
-    # we could do better, some guards might be redundant,
-    # e.g. Eq(s0, 4) & Eq(2*s0, 8)
-    # but unclear how to handle all of that right now.
-    # TODO(pianpwk): better way of doing this
-    new_ras = []
-    ras_exprs: Set[sympy.Expr] = set()
-    for ras in ras_by_symbol.pop(None, []):  # type: ignore[call-overload]
-        if ras.expr not in ras_exprs:
-            new_ras.append(ras)
-            ras_exprs.add(ras.expr)
-    ras_by_symbol[None] = new_ras  # type: ignore[index]
-
     # We are going to mutate the dict
     symbol_to_proxy: Dict[sympy.Symbol, fx.Proxy] = {}
     placeholders = set()
-    last_placeholder = None
+    first_non_placeholder = next(iter(graph.nodes))
     for node in graph.nodes:
         if node.op != "placeholder":
+            first_non_placeholder = node
             break
-        last_placeholder = node
-        placeholders.add(node)
-    if last_placeholder is None:  # no placeholders, just insert before first node
-        last_placeholder = next(iter(graph.nodes))
+        else:
+            placeholders.add(node)
+
+    def _is_intermediate_tensor_shape_call(node: fx.Node) -> bool:
+        """
+        If a size/stride/storage offset call on an intermediate tensor,
+        we can try to compute the value from input shapes instead.
+        """
+        val = None
+        if (
+            node.target in (
+                torch.ops.aten.sym_size.int,
+                torch.ops.aten.sym_stride.int,
+                torch.ops.aten.storage_offset.default,
+            )
```

Comment:
I actually think you can probably do some more generic test here. Specifically, if I have a function which returns a SymInt, and whose inputs contain a Tensor, THEN I should try to recompute from input shapes. This covers all of the functions here, but really there is nothing special about these functions, any function that returns a SymInt ought to be eligible.